<launch>
  <!-- Tuỳ chọn -->
  <arg name="world"           default="$(find agv_rio_final1)/worlds/empty.world"/>
  <arg name="start_rviz"      default="true"/>
  <arg name="start_teleop"    default="true"/>
  <arg name="start_mapping"   default="true"/>
  <arg name="use_fake_jsp"    default="false"/>  <!-- true nếu bạn KHÔNG có joint_states từ Gazebo -->

  <!-- Đồng bộ thời gian mô phỏng -->
  <param name="use_sim_time" value="true"/>

  <!-- 1) Gazebo + world -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(arg world)"/>
  </include>

  <!-- 2) Nạp URDF vào /robot_description -->
  <param name="robot_description"
         command="$(find xacro)/xacro '$(find agv_rio_final1)/urdf/agv_rio_final1.urdf'"/>

  <!-- 3) Spawn robot -->
  <node name="spawn_model" pkg="gazebo_ros" type="spawn_model"
        args="-urdf -param robot_description -model agv_rio_final1"
        output="screen"/>

  <!-- 4) TF & state publishers (để RViz vẽ đúng) -->
  <node name="tf_footprint_base" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 base_link base_footprint 40"/>
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen"/>
  <group if="$(arg use_fake_jsp)">
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" output="screen"/>
  </group>

  <!-- 5) SLAM (gmapping) -->
  <group if="$(arg start_mapping)">
    <node pkg="slam_gmapping" type="slam_gmapping" name="slam_gmapping" output="screen">
      <remap from="scan" to="/agv/scan"/>
      <param name="base_frame" value="base_link"/>
      <param name="odom_frame" value="odom"/>
      <param name="map_update_interval" value="2.0"/>
      <param name="linearUpdate"  value="0.20"/>
      <param name="angularUpdate" value="0.10"/>
      <param name="xmin" value="-20.0"/><param name="ymin" value="-20.0"/>
      <param name="xmax" value=" 20.0"/><param name="ymax" value=" 20.0"/>
      <param name="delta" value="0.05"/>
      <param name="maxUrange" value="11.5"/>
    </node>
  </group>

  <group if="$(arg start_rviz)">
   
    <node pkg="rviz" type="rviz" name="rviz"
          args="-d $(find agv_rio_final1)/rviz/agv_lidar.rviz"
          output="screen"/>
  </group>


  <group if="$(arg start_teleop)">
 
    <env name="SDL_VIDEODRIVER" value="x11"/>
    <node pkg="agv_rio_final1" type="pygame_wasd_teleop.py" name="wasd_teleop" output="screen">
      <param name="cmd_topic"     value="/agv/cmd_vel"/>
      <param name="rate_hz"       value="50"/>
      <param name="max_linear"    value="0.8"/>
      <param name="max_angular"   value="2.0"/>
      <param name="step_linear"   value="0.1"/>
      <param name="step_angular"  value="0.2"/>
    </node>
  </group>
</launch>
